### PR TITLE
Enhancement: detect public open port from network flows #40

### PR DESCRIPTION
**Pull request by blackfirespace**
Enhancement: detect public open port from network flows #40
Enhancement: Able to nmap router public ip to scan open ports #14

### CODE REVIEW

File: alarm/Alarm.js [hunk 1]
```diff
@@ -882,6 +929,20 @@ class UpnpAlarm extends Alarm {
...
     this["p.device.name"]];
   }
+  
+  isDup(alarm) {
+    if (alarm.type === this.type) {
+      return super.isDup(alarm);
+    };
+
+    const sameType = ['ALARM_OPENPORT', 'ALARM_UPNP'];
+    if (!(sameType.indexOf(alarm.type) !== -1 && sameType.indexOf(this.type) !== -1))
```
[the remark this answers is not in the document]
Comment (drahnieR) [reply alarm/Alarm.js]: type of `this` is always ALARM_UPNP

File: sensor/ExternalScanSensor.js [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+
+    return confirmedPorts;
+  }
+
+  _scan(publicIP) {
+    let cmd = util.format('sudo nmap -Pn -F %s -oX - | %s', publicIP, xml2jsonBinary);
+
+    log.info("Running command:", cmd);
+    return new Promise((resolve, reject) => {
+      cp.exec(cmd, (err, stdout, stderr) => {
```
Comment (drahnieR) [inline sensor/ExternalScanSensor.js]: use `child-process-promise`

File: sensor/ExternalScanSensor.js [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+      throw new Error(msg + '\n' + err.message)
+    }
+  }
+
+  async cloudConfirmOpenPort(publicIP, port) {
+    let result = false;
+    try {
+      const token = await tokenManager.getToken();
+      let options = {
+        uri: util.format("http://scan.encipher.io:9999/scan/lzk_cfwSa-DDCWlPURC5Tw/%s/%s/%s", publicIP, publicIP, port),
```
Comment (MelvinTo) [inline sensor/ExternalScanSensor.js]: this lzk part needs to be replaced with the local eid

File: sensor/ExternalScanSensor.js [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+        await rclient.expireAsync(redisKey, 86400);
+      } catch (err) {
+        log.error("Failed to scan external ports: " + err);
+      }
+    }
+
+    return host;
+  }
+
+  async rrWithErrHandling(options) {
```
Comment (MelvinTo) [inline sensor/ExternalScanSensor.js]: if this is so common used, should it be made as a common lib?

Comment (MelvinTo) [reply sensor/ExternalScanSensor.js]: @drahnieR

File: sensor/ExternalScanSensor.js [hunk 5]
```diff
@@ -159,41 +162,24 @@ class ExternalScanSensor extends Sensor {
...
+      log.info("Cloud confirm: ", uri);
       let options = {
-        uri: util.format("http://scan.encipher.io:9999/scan/lzk_cfwSa-DDCWlPURC5Tw/%s/%s/%s", publicIP, publicIP, port),
+        uri: uri,
         method: 'GET',
         auth: {
           bearer: token
         },
-        retryDelay: 1000,  // (default) wait for 1s before trying again
+        retryDelay: 3000,  // (default) wait for 1s before trying again
```
Comment (drahnieR) [inline sensor/ExternalScanSensor.js]: 3000 is not default 1s

File: sensor/ExternalScanSensor.js [hunk 6]
```diff
@@ -216,6 +202,7 @@ class ExternalScanSensor extends Sensor {
         if (isOpen) {
           confirmedPorts.push(current);
         }
+        await delay(5000);
```
Comment (drahnieR) [inline sensor/ExternalScanSensor.js]: why there is a delay, add some comments

**Pull request merged**